File: PAT_Basic_Level_by_cpp/1025 反转链表/1025 反转链表.cpp
// 1025 反转链表
// 给定链表首地址、结点总数 N 与正整数 K，将链表中每 K 个结点为一组进行反转，
// 最后不足 K 个的结点保持原序。输出反转后的链表（格式同输入）。
//
// 【实现原理】
// 算法思路：
//   1. 输入所有结点信息，以地址为索引存入数组 nodes[100000]（地址为 5 位非负整数，O(1) 查找）。
//   2. 从首地址开始，沿 next 指针遍历链表，按顺序收集有效结点的地址到 order 数组。
//      （注意：输入中可能存在不在链表上的结点，需要忽略。）
//   3. 对 order 数组进行分组反转：
//      - 每 K 个结点为一组，使用 reverse() 反转该组内的地址顺序。
//      - 最后不足 K 个的组保持原序。
//   4. 根据反转后的 order 顺序输出结果：
//      - 每个结点的 next 地址为 order 中下一个结点的地址（最后一个为 -1）。
// 关键点：
//   - 输入中可能存在多余结点（不在链表上），需要按链表顺序重新收集有效结点。
//   - 反转的是地址顺序，而非实际移动结点数据，高效且简单。
//   - 输出时地址格式为 5 位数字（不足补前导 0），-1 直接输出。
// 时间复杂度：O(N)，遍历链表和反转均为线性时间。
// 空间复杂度：O(N)，存储结点信息和地址顺序。
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Node {
    int addr;
    int data;
    int next;
};

int main() {
    int head, n, k;
    if (!(cin >> head >> n >> k)) return 0;

    // 用地址直接作下标，O(1) 查找（地址为 5 位非负整数）
    vector<Node> nodes(100000);
    for (int i = 0; i < n; ++i) {
        int addr, data, next;
        cin >> addr >> data >> next;
        nodes[addr] = {addr, data, next};
    }

    // 按链表顺序收集有效结点地址（忽略不在链上的多余结点）
    vector<int> order;
    int p = head;
    while (p != -1) {
        order.push_back(p);
        p = nodes[p].next;
    }

    int len = (int)order.size();
    // 每 K 个反转一组
    for (int i = 0; i + k <= len; i += k) {
        reverse(order.begin() + i, order.begin() + i + k);
    }

    // 输出，注意下一个地址需指向反转后的后继
    for (int i = 0; i < len; ++i) {
        int addr = order[i];
        int nextAddr = (i + 1 < len) ? order[i + 1] : -1;
        if (nextAddr == -1)
            printf("%05d %d -1\n", addr, nodes[addr].data);
        else
            printf("%05d %d %05d\n", addr, nodes[addr].data, nextAddr);
    }
    return 0;
}

File: PAT_Basic_Level_by_go/1025 反转链表/1025 反转链表.go
/*
 * 题目：1025 反转链表
 * 实现原理：用 map 存储所有节点（地址→节点），从给定首地址开始遍历链表，将节点按顺序
 * 存入切片。然后每 K 个节点为一组进行反转（原地交换），最后更新每个节点的 next 指针
 * 指向下一节点的地址并输出。
 */
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type node struct {
	addr int
	data int
	next int
}

// 反转链表：每 K 个节点为一组反转，输出新链表。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	if !scanner.Scan() {
		return
	}
	f := strings.Fields(scanner.Text())
	head, _ := strconv.Atoi(f[0])
	N, _ := strconv.Atoi(f[1])
	K, _ := strconv.Atoi(f[2])
	nodes := make(map[int]node) // 用 map 以地址为键存储所有节点
	for i := 0; i < N; i++ {
		if !scanner.Scan() {
			break
		}
		p := strings.Fields(scanner.Text())
		a, _ := strconv.Atoi(p[0])
		d, _ := strconv.Atoi(p[1])
		nx, _ := strconv.Atoi(p[2])
		nodes[a] = node{addr: a, data: d, next: nx}
	}
	list := make([]node, 0)
	// 从首地址开始，按 next 指针遍历，将有效节点存入切片
	for addr := head; addr != -1; addr = nodes[addr].next {
		n, ok := nodes[addr]
		if !ok {
			break
		}
		list = append(list, n)
	}
	// 每 K 个节点为一组，原地反转
	for i := 0; i+K <= len(list); i += K {
		for l, r := i, i+K-1; l < r; l, r = l+1, r-1 {
			list[l], list[r] = list[r], list[l]
		}
	}
	// 输出结果，每个节点的 next 指向下一节点的地址
	for i := 0; i < len(list); i++ {
		nx := -1
		if i+1 < len(list) {
			nx = list[i+1].addr
		}
		fmt.Printf("%05d %d %s\n", list[i].addr, list[i].data, formatAddr(nx))
	}
}

func formatAddr(addr int) string {
	if addr == -1 {
		return "-1"
	}
	return fmt.Sprintf("%05d", addr)
}
